src/semantic_world/geometry.py

```diff
@@ -476,33 +456,88 @@ def get_points(self) -> List[Point3]:
                 for z in (self.min_z, self.max_z)]
 
     @classmethod
-    def from_min_max(cls, min_point: Point3, max_point: Point3) -> Self:
+    def from_min_max(
+        cls, reference_frame: Body, min_point: Point3, max_point: Point3
+    ) -> Self:
         """
         Set the axis-aligned bounding box from a minimum and maximum point.
 
         :param min_point: The minimum point
         :param max_point: The maximum point
         """
-        return cls(*min_point.to_np()[:3], *max_point.to_np()[:3])
+        return cls(
+            *min_point.to_np()[:3],
+            *max_point.to_np()[:3],
+            reference_frame=reference_frame,
+        )
 
-    def as_shape(self, reference_frame: Optional[Body] = None) -> Box:
+    def as_shape(self) -> Box:
         scale = Scale(x=self.max_x - self.min_x,
                       y=self.max_y - self.min_y,
                       z=self.max_z - self.min_z)
         x = (self.max_x + self.min_x) / 2
         y = (self.max_y + self.min_y) / 2
         z = (self.max_z + self.min_z) / 2
-        origin = TransformationMatrix.from_xyz_rpy(x, y, z, 0, 0, 0, reference_frame)
+        origin = TransformationMatrix.from_xyz_rpy(
+            x, y, z, 0, 0, 0, self.reference_frame
+        )
         return Box(origin=origin, scale=scale)
 
+    def transform_to_frame(self, reference_frame: Body):
+        """
+        Returns the bounding box of the box.
+        """
+
+        world = self.reference_frame._world
+        origin_frame = self.reference_frame
+        reference_T_origin = world.compute_forward_kinematics(
+            reference_frame, origin_frame
+        )
+
+        # Get all 8 corners of the BB in link-local space
+        list_origin_T_corner = [
+            TransformationMatrix.from_point_rotation_matrix(origin_P_corner)
+            for origin_P_corner in self.get_points()
+        ]  # shape (8, 3)
+
+        list_reference_T_corner = [
+            reference_T_origin @ origin_T_corner
+            for origin_T_corner in list_origin_T_corner
+        ]
+
+        list_reference_P_corner = [
+            reference_T_corner.to_position().to_np()[:3]
+            for reference_T_corner in list_reference_T_corner
+        ]
+
+        # Compute world-space bounding box from transformed corners
+        min_corner = np.min(list_reference_P_corner, axis=0)
+        max_corner = np.max(list_reference_P_corner, axis=0)
+
+        world_bb = BoundingBox.from_min_max(
+            reference_frame,
+            Point3.from_iterable(min_corner),
+            Point3.from_iterable(max_corner),
+        )
+
+        return world_bb
+
 
 @dataclass
 class BoundingBoxCollection:
     """
     Dataclass for storing a collection of bounding boxes.
     """
 
+    reference_frame: Body
+    """
+    The reference frame of the bounding boxes.
+    """
+
```

Comment:
i don't think a bounding box collection needs a reference frame. 
the bounding boxes already have one, so what would the frame of a collection mean?

Reply:
if it's the frame that all bounding boxes should have, then make sure in __post_init__ that they do and give the collection a property instead, if you need.